course/page/code.py

```diff
@@ -1190,4 +1190,945 @@ def grade(self, page_context, page_data, answer_data, grade_data):
 
 # }}}
 
```

Comment:
Throughout this file, there is far too much code duplication for my liking. For example:

* I would much prefer if both code questions inherited from a common base class.
* Why are there two code forms?
* Why are there two copies of `request_*_run`?

Reply:
To explain the rationale:

* Duplicated code is very difficult to review: The changes are not obvious.
* Duplicated code is even harder to maintain---a cost that I'm reluctant to shoulder.